pages/blog/graphql-codegen-and-apollo-client-3.mdx

```diff
@@ -0,0 +1,176 @@
+
+Then, when you create your Apollo Client cache instance, use the generated variable:
+
+```ts
+import introspection from './src/possible-types';
+
+export default new ApolloClient({
+  uri: 'https://countries.trevorblades.com',
+  cache: new InMemoryCache({ possibleTypes: introspection.possibleTypes }),
+});
+```
+
+## Type Policies
+
+If you are an [advanced configuration for your Apollo-Client cache](https://www.apollographql.com/docs/react/caching/cache-configuration/#data-normalization), you can customize the behaviour of your cache.
+
+The configuration you pass to Apollo depends on your GraphQL types and their fields, and instead of having an arbitrary object, you can have a fully-typed signature generated based on your GraphQL schema.
+
+You can use `@graphql-codegen/typescript-apollo-client-helpers` plugin to generate that.
+
+```yaml
+schema: my-schema.graphql
+generates:
+  ./src/type-policies.ts:
+    plugins:
+      - typescript-apollo-client-helpers
+```
+
+Then, use the generated `TypedTypePolicies` to type your object:
+
+```ts
+import { TypedTypePolicies } from './apollo-helpers';
+
+const typePolicies: TypedTypePolicies = {
+  // Keys in this object will be validated against the typed on your schema
+  Product: {
+    keyFields: ['id'], // Values in this field will be validated against the available fields from the Product type
+  },
+  Person: {
+    keyFields: ['name', 'email'],
+  },
+  Book: {
+    // This entire definition is typed, based on available types and fields
+    fields: {
+      tags: {
+        merge: false,
+      },
+    },
+  },
+};
+
+const cache = new InMemoryCache({
+  typePolicies,
+});
+```
+
+## TypedDocumentNode
+
+Apollo-Client also supports `TypedDocumentNode` now (since `v3.2`, [you can read more about it here](https://the-guild.dev/blog/typed-document-node)).
+
+You can use it to generate a fully-typed `DocumentNode` you can use with Apollo-Client, and it will automatically type your variables and responses.
+
+You can use `@graphql-codegen/typed-document-node` with the following setup to get that:
+
+```yaml
+schema: schema.graphql
+documents: query.graphql
+generates:
+  ./typed-document-nodes.ts:
+    plugins:
+      - typescript
+      - typescript-operations
+      - typed-document-node
+```
+
+Later, in your code, you can just import the generated `TypedDocumentNode` objects from `typed-document-nodes` (based on your GraphQL operations), and it will be automatically typed:
+
+```tsx
+import { RatesDocument } from './typed-document-nodes';
+import { useQuery } from '@apollo/client';
+
+export const MyComponent: React.FC = () => {
+  // We now have types support and auto complete for the
+  // result type, just by passing `RatesDocument` as `query` to apollo client.
+  const result = useQuery(RatesDocument, {
+    variables: {
+      currency: 'USD',
+    },
+  });
+
+  const rates = result.data.rates;
+
+  return <div>Rates: {rates}</div>;
+};
+```
+
+## Ready-to-use Hooks/HOC/Components
```

Comment:
Should we also mention `graphql-let` somewhere?
How does `graphql-let` fits into all of this if any?

Reply:
It's a wrapper for codegen + webpack, I think it might not be so relevant here, it isn't specific for AC3...